pkgs/development/python-modules/remotecv/install_requires.patch

```diff
@@ -0,0 +1,16 @@
+diff --git a/setup.py b/setup.py
+index 70f765c..8003cda 100644
+--- a/setup.py
++++ b/setup.py
+@@ -53,9 +53,8 @@ remotecv is an OpenCV worker for facial and feature recognition
+     },
+ 
+     install_requires=[
+-        "argparse>=1.2.1,<1.3.0",
+-        "pyres>=1.5,<1.6",
+-        "Pillow>=4.3.0,<5.2.0",
++        "pyres>=1.5",
++        "Pillow>=4.3.0",
```

Comment:
I'm not sure if we want version constraints without upper bounds here. When major bumps happen (and `remotecv` isn't compatible with them) we may end up having succeeding packages with broken build products.

I recommend something like `pyres>=1.5,<2.0`.

Reply:
Thanks for the comment! I think I disagree though, if I'm understanding correctly. Please don't hesitate to disagree with me.

First, regarding remotecv itself: In my opinion using version upper bounds "just in case" is a bad practice. Remotecv needs to do a new release every time there's a new release of any of its dependencies it has upper bounded. If every package had that kind of policy, the whole Python ecosystem would be practically broken because of a need of recursive new releases. Of course, if the version is incompatible, then yes, needs an upper bound. Someone has written an issue about this: https://github.com/thumbor/remotecv/issues/15

Second, regarding our packaging: I'm not sure why we should replace the upper bounds written by Remotecv itself with something else instead of removing them. If the package creator has set 1.6 upper bound, why would we put 2.0 instead of just removing it? We have the check phase to make sure it works. If we set our own upper bound, we need to modify this patch whenever pyres or pillow is upgraded on NixOS. We aren't setting our own version upper bounds to any other packages just in case. That would cause recursive modification of package definitions whenever a dependency is upgraded. Or is there some particular reason that with remotecv and pyres/pillow this would be a good practice even if it wouldn't be good in general?

Reply:
> First, regarding remotecv itself: In my opinion using version upper bounds "just in case" is a bad practice. Remotecv needs to do a new release every time there's a new release of any of its dependencies it has upper bounded. If every package had that kind of policy, the whole Python ecosystem would be practically broken because of a need of recursive new releases. Of course, if the version is incompatible, then yes, needs an upper bound.

Your argument certainly applies for "minor" upper bounds (e.g. `>=1.2,<1.3`). But I'm talking about disallowing the next major version (e.g. `>=1.2,<2.0`). Major releases are expected to have API incompatibilities (as of SemVer), so I don't see why restricting them is an issue. I absolutely agree with you though that it's bad idea to have upper bounds as used in upstream `remotecv` though, in such a case you're absolutely right, one would have to update the package after each minor (non-breaking) release.

> Second, regarding our packaging: I'm not sure why we should replace the upper bounds written by Remotecv itself with something else instead of removing them. If the package creator has set 1.6 upper bound, why would we put 2.0 instead of just removing it? We have the check phase to make sure it works. If we set our own upper bound, we need to modify this patch whenever pyres or pillow is upgraded on NixOS. We aren't setting our own version upper bounds to any other packages just in case.

I think that restricting `2.0` is the most sensitive variant. Your patch allows e.g. `pyres` 3.0 ATM as well.

Let's say folks don't upgrade `pyres` in the next months and upstream releases 3.0. A bump happens in `nixpkgs` and `remotecv` still builds **although** it's incompatible with `pyres` 3.0 (e.g. since 3.0 support isn't released yet). Now we have a working build, but the result can be broken at runtime due to API changes.

I prefer to actually have breaking packages when doing a major bump (these can be covered easily with `nix-review`) rather than weird runtime failures. Also it's not clear if a simple update of the version constraint is sufficient, you may have to upgrade this package as well to get it working (or maintain two major versions of $package).

I hope that makes my concerns a bit clearer :)

Reply:
I see your concern but I still slightly disagree: I think unit tests should check if it'll work runtime or not. Working build should thus mean working runtime compatibility with the dependencies. If unit tests don't capture runtime failures in case of dependency incompatibility, I think it's a problem in the unit tests. So, if `remotecv` builds and checks fine, then it should work runtime with the given dependencies. And this is all done automatically in nixpkgs, thus I don't see any reason to add this kind of manual restrictions unless there are some particular concerns on some package dependencies or the upstream package has explicitly made those upper bounds for some good reason.

`nox-review` will also build recursively and run check phases this so I don't see the problem with that either.. Is your concern that the unit test coverage on `remotecv` is bad and we should thus protect these particular dependency upgrades? For which other packages should we add this kind of upper bound restrictions too? But I can add that change if wanted, no problem, I just don't think it's good practice as far as I understand.

Reply:
Ah right that's clearly my mistake, I commented about some `checkPhase` things earlier and I had in mind that the patch was for this package.

That's clearly my mistake, sorry for the noise!

Reply:
Ah, ok!